bundles/org.eclipse.text/src/org/eclipse/jface/text/source/AnnotationModel.java

```diff
@@ -789,15 +799,24 @@ public void removeAllAnnotations() {
 	protected void removeAllAnnotations(boolean fireModelChanged) {
 		IAnnotationMap annotations= getAnnotationMap();
 		if (fDocument != null) {
-			Iterator<Annotation> e= getAnnotationMap().keySetIterator();
-			while (e.hasNext()) {
-				Annotation a= e.next();
-				Position p= annotations.get(a);
-				removePosition(fDocument, p);
-//				p.delete();
-				synchronized (getLockObject()) {
-					getAnnotationModelEvent().annotationRemoved(a, p);
+			AnnotationModelEvent event;
+			synchronized (getLockObject()) {
+				event = getAnnotationModelEvent();
+			}
+			if (!(annotations instanceof IAnnotationMap2)) {
+				Iterator<Annotation> e= getAnnotationMap().keySetIterator();
+				while (e.hasNext()) {
+					Annotation a= e.next();
+					Position p= annotations.get(a);
+					removePosition(fDocument, p);
+					event.annotationRemoved(a, p);
 				}
+			} else {
+				// foreach is synchronized on map lock object
+				annotations.forEach((a, p) -> {
+					removePosition(fDocument, p);
```

Comment:
@szarnekow: I think this line is the last reason I would keep on new Interface. The `removePosition()` can be overridden, and use extra locks / operations that might interfere with the new enclosing lock in the map.forEach().

Reply:
I agree.

Reply:
One more thought: Since the class `AnnotationMap` is package private, a check for that specific class might also do the trick.

Reply:
But. of course that would harm extensibility

Reply:
> check for that specific class might also do the trick

Jawoll!

Reply:
```
The removePosition() can be overridden, and use extra locks / operations that might interfere with the new enclosing lock in the map.forEach().
```

Maybe I don't agree.
Let's assume I've a subtype of AnnotationModel which uses the regular AnnotationMap - it now implements IAnnotationMap2.

How does the instanceof IAnnotationMap2 help me with my newly introduced locks in my overridden AnnotationModelSubtype?

Reply:
> How does the instanceof help me with my newly introduced locks in my overridden AnnotationModelSubtype?

Doesn't help for existing code, right. I will check if we can skip changes on removeAllAnnotations method for now.

Reply:
OK, simplified the patch as much as possible to avoid any side effects.